src/passes/call-utils.h

```diff
@@ -0,0 +1,153 @@
+// form:
+//
+//  (if
+//    (..condition..)
+//    (call $A)
+//    (call $B)
+//  )
+//
+// If we fail to find the expected pattern, or we decide it is not worth
+// optimizing it for some reason, we return nullptr.
+//
+// If this returns the new form, it will modify the function as necessary,
+// adding new locals etc., which later passes should optimize.
+//
+// |getCallInfo| is given one of the arms of the select and should return an
+// IndirectCallInfo that says what we know about it. We may know nothing, or
+// that it will trap, or that it will go to a known static target.
+template<typename T>
+inline Expression*
+convertToDirectCalls(T* curr,
+                     std::function<IndirectCallInfo(Expression*)> getCallInfo,
+                     Function& func,
+                     Module& wasm) {
+  auto* select = curr->target->template dynCast<Select>();
+  if (!select) {
+    return nullptr;
+  }
+
+  if (select->condition->type == Type::unreachable) {
+    // Leave this for DCE.
+    return nullptr;
+  }
+
+  // Check if we can find useful info for both arms: either known call targets,
+  // or traps.
+  // TODO: support more than 2 targets (with nested selects)
+  auto ifTrueCallInfo = getCallInfo(select->ifTrue);
+  auto ifFalseCallInfo = getCallInfo(select->ifFalse);
+  if (std::get_if<Unknown>(&ifTrueCallInfo) ||
+      std::get_if<Unknown>(&ifFalseCallInfo)) {
+    // We know nothing about at least one arm, so give up.
+    // TODO: Perhaps emitting a direct call for one arm is enough even if the
+    //       other remains indirect?
+    return nullptr;
+  }
+
+  auto& operands = curr->operands;
+
+  // We must use the operands twice, and also must move the condition to
+  // execute first; use locals for them all. While doing so, if we see
+  // any are unreachable, stop trying to optimize and leave this for DCE.
+  std::vector<Index> operandLocals;
+  for (auto* operand : operands) {
+    if (operand->type == Type::unreachable ||
+        !TypeUpdating::canHandleAsLocal(operand->type)) {
+      return nullptr;
+    }
+  }
+
+  Builder builder(wasm);
+  std::vector<Expression*> blockContents;
+
+  // None of the types are a problem, so we can proceed to add new vars as
+  // needed and perform this optimization.
+  for (auto* operand : operands) {
+    auto currLocal = builder.addVar(&func, operand->type);
+    operandLocals.push_back(currLocal);
+    blockContents.push_back(builder.makeLocalSet(currLocal, operand));
+  }
+
+  // Build the calls.
+  auto numOperands = operands.size();
+  auto getOperands = [&]() {
+    std::vector<Expression*> newOperands(numOperands);
+    for (Index i = 0; i < numOperands; i++) {
+      newOperands[i] =
+        builder.makeLocalGet(operandLocals[i], operands[i]->type);
+    }
+    return newOperands;
+  };
+
+  auto makeCall = [&](IndirectCallInfo info) -> Expression* {
+    if (std::get_if<Trap>(&info)) {
+      return builder.makeUnreachable();
```

Comment:
Why not handle `Trap` earlier where we handled `Unknown`?

Reply:
If we checked for `Trap` where we check for `Unknown`, could we either replace the entire expression with `unreachable` at that point or skip the optimization to let DCE do that? That would save us the work of going through and doing the rest of the optimization.

Reply:
Oh, the trap might only happen in one arm:
```wat
(call_indirect
  (select
    (i32.const 1) ;; goes to $valid_target
    (i32.const 99999) ;; invalid
    (..condition..)))

=>

(if
  (..condition..)
  (call $valid_target)
  (unreachable))
```
We have a [test for that](https://github.com/WebAssembly/binaryen/blob/ac2190c1bd1a0be12a1d25e62695f791f270d050/test/lit/passes/directize_all-features.wast#L619-L659) (existing before this PR).

Reply:
oh I see, thanks!